source/extensions/filters/http/ext_proc/client_impl.cc

```diff
@@ -75,12 +79,18 @@ void ExternalProcessorStreamImpl::onReceiveTrailingMetadata(Http::ResponseTraile
 void ExternalProcessorStreamImpl::onRemoteClose(Grpc::Status::GrpcStatus status,
                                                 const std::string& message) {
   ENVOY_LOG(debug, "gRPC stream closed remotely with status {}: {}", status, message);
-  callbacks_.logGrpcStreamInfo();
   stream_closed_ = true;
+
+  if (!callbacks_.has_value()) {
+    ENVOY_LOG(debug, "Underlying filter object has been destroyed.");
```

Comment:
I would think that in this case the stream should delete itself, since there is no point to keep it around.

Reply:
The stream is remotely closed  **after** this callback invocation  [definition here](https://github.com/envoyproxy/envoy/blob/main/envoy/grpc/async_client.h#L142-L145) .  Thus, this callback can still be invoked normally before stream is fully deleted.

Actually I have integration test with tsan to reproduce this dangling reference issue without this check

Reply:
I understand this check here. The issue as I understand the code is that the deferred stream is deleted only by the expired timer, even though the remote server closed it before the timer has expired. If the timer is very long, you can have these objects accumulating in the memory even though they no longer needed.

Reply:
My understanding is that `localClose` and `remoteClosebySever` are independent. stream does delete itself in this case and it is not affected by change here.

Without this change, when sever remotely close the stream, local stream object won't be deleted (as it has already been closed) ---- stream_closed is set to true onRemoteClose and [localClose](https://github.com/envoyproxy/envoy/blob/69bb97b98354b6946c1fd62f9c1ca641d0080dcf/source/extensions/filters/http/ext_proc/client_impl.cc#L57-L63) is a no-op . The `stream_ pointer` is not deleted until the filter onDestroy but its associated resource have been freed up .

With this change, it is same that stream's associated resource have already been freed up, which is also why `stream_closed_ ` is set to true here to avoid double-deletion.  The only difference here is that `stream_ pointer` is not deleted in filter onDestroy but is deleted when timer expired.

So IIUC, the PR doesn't validate your initial point that stream should delete itself and not being kept around.

Reply:
Per offline discussion, early deletion of `stream_ pointer` onRemoteClose will be explored in the next PR, as the performance optimization